File: Graphs/LC_127_Word_Ladder.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.18

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordlist = set(wordList)
        q = collections.deque()
        q.append(beginWord)
        steps = 0
        if endWord not in wordlist:
            return 0
        
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                if cur == endWord:
                    return steps + 1
                for i in range(len(cur)):
                    for ch in "abcdefghijklmnopqrstuvwxyz":
                        if ch == cur[i]:
                            continue
                        temp = cur[:i] + ch + cur[i+1:]
                        if temp in wordlist:
                            q.append(temp)
                            wordlist.remove(temp)
            steps += 1
        
        return 0
                     

File: Graphs/LC_127_Word_Ladder.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.18

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordlist(wordList.begin(), wordList.end());
        if (!wordlist.count(endWord)){
            return false;
        }
        queue<string> q;
        q.push(beginWord);
        int steps = 0;
        while (!q.empty()){
            int size = q.size();
            for (int n = 0; n < size; n++){
                string cur = q.front();
                q.pop();
                if (cur == endWord){
                    return steps + 1;
                }
                for(int i = 0; i < cur.size(); i++){
                    char c = cur[i];
                    for(int j = 0; j < 26; j++){
                        cur[i] = 'a' + j;
                        if (wordlist.count(cur)){
                            q.push(cur);
                            wordlist.erase(cur);
                        }
                    }
                    cur[i] = c;
                }
            }
            steps++;
        }

        return 0;   
    }
};